torch/csrc/autograd/functions/accumulate_grad.h

```diff
@@ -11,7 +11,29 @@ struct TORCH_API AccumulateGrad : public Node {
 
   variable_list apply(variable_list&& grads) override;
 
+  // Given a variable with its current grad as variable_grad, accumulates
+  // new_grad into variable_grad if in place accumulation is possible.
+  // Otherwise, uses 'update_grad' to update the grad for the variable.
+
+  // variable: the variable whose grad we're accumulating.
+  // variable_grad: the current grad for the variable.
+  // new_grad: new grad we want to acummulate for the variable.
+  // num_expected_refs: the number of refs we expect to hold internally
+  //                    such that it is safe to avoid cloning the grad
+  //                    if use_count() of the grad is less than or equal
+  //                    to this value (in addition to post_hooks).
+  // update_grad: Function that is used to update grad for the variable.
+  //              The argument to the function is a Tensor which
+  //              is used to set a new value for the grad.
+  static void accumulateGradAndCallHooks(
+      const Variable& variable,
+      at::Tensor variable_grad,
+      const at::Tensor& new_grad,
+      size_t num_expected_refs,
+      const std::function<void(at::Tensor&&)>& update_grad);
```

Comment:
`std::function` here looks slow, because we have to convert the lambda object into the `std::function` representation. Example https://github.com/facebook/folly/issues/344

It seems to me that you just wanted to reuse code here, and effectively just inline `accumulateGradAndCallHooks` into the call sites. You can do that if you convert this into a template function, templated over the (unnamed) lambda type, then you can have your cake and eat it too.